### PR TITLE
Add filter and recall autocomplete options

### LINKED ISSUE #12783: add option to refresh auto-complete list instead of jumping
### Description of the Issue

When you have the auto-complete number lowered to 1 or 2 (which is often useful), the list of options can get very long, so the current behaviour of jumping to different portions instead of refreshing/trimming the list can be cumbersome (especially since the list is grouped by category).

### Steps to Reproduce the Issue

1. Add the following strings to a fresh file: abcd, abdc, acbd, acdb, adbc, adcb
2. Set the Auto-Complete number to 1.
3. Type 'a' to trigger an auto-complete list.
autocomplete1
5. Type 'd'.

### Current Behavior

The first 4 items remain, and the highlighted entry jumps to the 5th.

autocomplete2

### Requested Behavior

The first 4 items are cleared away.

autocomplete3

### Current Workaround
Delete the 'd' and retype.

### PR DESCRIPTION
**Pull request by mpheath**
Close #12783

### CODE REVIEW

**Comment by donho** [comment]
@mpheath 
Why do we need "Filter the list item" for "Function completion"?
Shouldn't "Function completion" works as "Word completion"?

**Comment by mpheath** [comment]
@donho 

>Why do we need "Filter the list item" for "Function completion"?

The previous behaviour for "Function completion" is to add all items from the xml file uinfiltered.
New behaviour is same though if "Filter the list items" is checked, then the characters typed filter the items to only those that match. The "Function completion" filtered is like the other completion options that always filter.

> Shouldn't "Function completion" works as "Word completion"?

If you are implying that "Function completion" should always filter, then some items in the xml may not ever be the list.

Example, ActionScript:

```actionscript
"allowscale"
"exec"
"fullscreen"
"quit"
"showmenu"
"trapallkeys"
#include
```
I guess are quoted constants and a preprocessor directive.

Languages with special characters starting with

 * AutoIt `@`, `#`
 * BaanC `#`
 * C `#`
 * Coffee `=>`, `->`
 ...
 ...

So, keeping the old "Function completion" behaviour can retain the ability to access these items that start with special characters.

With "Filter the list items", only items that start with alphanumerical and underscores are in the list. The Environmental option of `additionalWordChar` seems to have no effect as AutoIt has set `@` and yet typing `@` does not invoke the autocomplete.

If "Function completion" works as "Word completion" as in always filtered, then the special items above become inaccessable.

Perhaps `additionalWordChar` is bugged? If fixed then the idea of "Function completion" always filter might be closer to being acceptable.

Personally, I am no fan of "Function completion" without filtering, though some users may rely on it's behaviour to show items that a filtered list will not show. The "Filter the list items" does stand out like a workaround and would be look nicer without the checkbox, though their does seem like something is needed to make a filtered "Function completion" to be fully workable with support of starting with special characters.

Removal of the "Filter the list items" checkbox seems like a good idea if the matching of special characters could be improved later. I could give that :+1: . I just try not to upset others by removing their beloved existing options/features, though if considered OK, then it can be done.

<details>

<summary>Compare to SciTE</summary>

SciTE is like "Function completion" filtered by default. "Word completion" is invoked by <kbd>Ctrl</kbd>+<kbd>Enter</kbd> .

Language set to AutoIt.

Type `#` and autocomplete will show `#include`, ... in the list.
Type `@` and autocomplete will show `@AppDataCommonDir`, ... in the list.

These special start characters are in the property [`autocomplete.lexer.ignorecase`](https://www.scintilla.org/SciTEDoc.html#property-autocomplete.*.start.characters) which I have set to `$(chars.alpha)$(chars.numeric)_@#$` .

SciTE vs Notepad++ makes me think that Notepad++ `additionalWordChar` is bugged or is not like SciTE `autocomplete.lexer.ignorecase` . If  `additionalWordChar` is not for autocomplete start character, then something for it might be useful.

</details>

**Comment by ArkadiuszMichalski** [comment]
@mpheath 
>Perhaps additionalWordChar is bugged? If fixed then the idea of "Function completion" always filter might be closer to being acceptable.

`additionalWordChar` is used for another purpose: https://github.com/notepad-plus-plus/notepad-plus-plus/issues/10016#issuecomment-1312556987.

**Comment by mpheath** [comment]
<details>

<summary>This patch gets the autocomplete working with special characters.</summary>


``` diff
diff --git a/PowerEditor/src/ScintillaComponent/AutoCompletion.cpp b/PowerEditor/src/ScintillaComponent/AutoCompletion.cpp
index 10bb582d..49246695 100644
--- a/PowerEditor/src/ScintillaComponent/AutoCompletion.cpp
+++ b/PowerEditor/src/ScintillaComponent/AutoCompletion.cpp
@@ -287,6 +287,24 @@ static void sortInsensitive(vector<generic_string> &wordArray)
 	);
 }
 
+intptr_t AutoCompletion::getAutocWordLengthFromCaret()
+{
+	generic_string autocWordChars = L"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_@#$";
+
+	intptr_t curPos = _pEditView->execute(SCI_GETCURLINE);
+	if (curPos == 0)
+		return false;
+
+	char curLine[curPos + 1];
+	_pEditView->execute(SCI_GETCURLINE, curPos + 1, reinterpret_cast<LPARAM>(&curLine));
+
+	intptr_t startPos = curPos;
+	while ((startPos > 0) && (autocWordChars.find(curLine[startPos - 1]) != std::string::npos))
+		startPos--;
+
+	return curPos - startPos;
+}
+
 
 bool AutoCompletion::showAutoComplete(AutocompleteType autocType, bool autoInsert)
 {
@@ -295,8 +313,12 @@ bool AutoCompletion::showAutoComplete(AutocompleteType autocType, bool autoInser
 
 	// Get beginning of word and complete word
 
+	intptr_t wordSize = getAutocWordLengthFromCaret();
+	if (wordSize <= 0)
+		return false;
+
 	intptr_t curPos = _pEditView->execute(SCI_GETCURRENTPOS);
-	intptr_t startPos = _pEditView->execute(SCI_WORDSTARTPOSITION, curPos, true);
+	intptr_t startPos = curPos - wordSize;
 
 	if (curPos == startPos)
 		return false;
@@ -316,7 +338,7 @@ bool AutoCompletion::showAutoComplete(AutocompleteType autocType, bool autoInser
 		if (len >= bufSize)
 			return false;
 
-		intptr_t endPos = _pEditView->execute(SCI_WORDENDPOSITION, curPos, true);
+		intptr_t endPos = curPos;
 
 		size_t lena = (endPos > startPos)?(endPos - startPos):(startPos - endPos);
 		if (lena >= bufSize)
@@ -1013,11 +1035,9 @@ void AutoCompletion::update(int character)
 	if (!nppGUI._autocRecall && _pEditView->execute(SCI_AUTOCACTIVE) != 0)
 		return;
 
-	const int wordSize = 64;
-	TCHAR s[wordSize];
-	_pEditView->getWordToCurrentPos(s, wordSize);
+	intptr_t wordSize = getAutocWordLengthFromCaret();
 
-	if (lstrlen(s) >= int(nppGUI._autocFromLen))
+	if (wordSize >= int(nppGUI._autocFromLen))
 	{
 		if (nppGUI._autocStatus == nppGUI.autoc_word)
 			showWordComplete(false);
diff --git a/PowerEditor/src/ScintillaComponent/AutoCompletion.h b/PowerEditor/src/ScintillaComponent/AutoCompletion.h
index da236e7a..7be7b028 100644
--- a/PowerEditor/src/ScintillaComponent/AutoCompletion.h
+++ b/PowerEditor/src/ScintillaComponent/AutoCompletion.h
@@ -78,6 +78,7 @@ public:
 	// Autocomplete from path.
 	void showPathCompletion();
 
+    intptr_t getAutocWordLengthFromCaret();
 	void insertMatchedChars(int character, const MatchedPairConf & matchedPairConf);
 	void update(int character);
 	void callTipClick(size_t direction);

```
</details>

It uses it's own word character set. Using the default word character set messes up highlighting... so is bad to change. I see ScintillaEditView.h messing with `SCI_SETWORDCHARS` which aint good and does not affect the autocomplete for some reason.

`autocWordChars` should probably get the character set from the xml, else use a default character set if not set in the xml.

@ArkadiuszMichalski 
> `additionalWordChar` is used for another purpose

Thanks, most are for calltips I notice.

>How can we get a hint for #include? When we press # nothing happens.

Been working on it. That is what the patch does, though still need xml load/save code done... It is an early preview at the moment.

>Such a new setting in Notepad++ for each language (in xml autocomplete file) would be ideal.

Should be possible.

If I get this all done and working, this PR or next?

**Comment by ArkadiuszMichalski** [comment]
@donho 
>Why do we need "Filter the list item" for "Function completion"?
Shouldn't "Function completion" works as "Word completion"?

Actually "Function completion" doesn't works as "Word completion" in this aspect.

`Function completion` << fetches all words from xml and displays them in a list (regardless of what we type).
image
This allows us to manually select those that can't be selected, such as `#include`.

`Word completion` << fetches and display in a list only those words that matches to what we type
image

`Function and word completion` <<  fetches from xml/document and display in a list those words that matches to what we type (so it's similar to above Word completion)
image

Is there any advantage to displaying all possible autocomplete  words from xml? Probably it's useful to view all the autocomplete words without opening the xml file manually.

It seems to me that the problem with suggesting words like `#include` (those with special characters) should be fixed first, and then decide what to do with this filtering for `Function completion`.

**Comment by ArkadiuszMichalski** [comment]
@mpheath 

On this PR https://github.com/notepad-plus-plus/notepad-plus-plus/pull/3198 someone dynamically extend Word List before showing autocomplete list and restore original state after closing it, so it doesn't affect highlighting.

>Should be possible.
>If I get this all done and working, this PR or next?

I think it should be in a separate PR, regardless of this one. There are some issues opened for problem with autocomplete where words containing special characters not work.

It's best to firstly determine how to specify these characters, whether using `additionalWordChar` or some new attribute in xml file.

**Comment by mpheath** [comment]
@ArkadiuszMichalski 

> It's best to firstly determine how to specify these characters, whether using `additionalWordChar` or some new attribute in xml file.

`additionalWordChar` seems good. I was going to add another attribute name, though if this is useful for the same names to get autocomplete (and calltip) words, then OK.

python.xml sets `.` which helps the autocomplete to recognize words with `.` .
lua.xml `.:` , cs.xml `.` , cobol `-` , BaanC.xml `.:$#` , autoit.xml `@`

All seem OK to use.

Before fix in lua, type `table.con`, close autocomplete, type `c` and no autocomplete shows. After fix, autocomplete shows as `.` is recognized as part of the word. So this will help with using recall...

**Review by donho — CHANGES_REQUESTED** [review]
1. please follow the suggestion.
2. If I'm not missing anything, we can remove "Filter the list item" under "Function completion", and it should remove - When user enable `Make auto-completion list brief` option, the function list is also filtered. Let's talk about it here if I'm missing somethings

File: PowerEditor/src/WinControls/Preference/preference.rc [hunk 1]
```diff
@@ -389,25 +390,26 @@ BEGIN
...
-    RTEXT           "Matched pair 1:",IDC_MACHEDPAIR_STATIC1,195,122,70,8
-    EDITTEXT        IDC_MACHEDPAIROPEN_EDIT1,267,120,14,14
-    EDITTEXT        IDC_MACHEDPAIRCLOSE_EDIT1,295,120,14,14
-    RTEXT           "Matched pair 2:",IDC_MACHEDPAIR_STATIC2,195,142,70,8
-    EDITTEXT        IDC_MACHEDPAIROPEN_EDIT2,267,140,14,14
-    EDITTEXT        IDC_MACHEDPAIRCLOSE_EDIT2,295,140,14,14
-    RTEXT           "Matched pair 3:",IDC_MACHEDPAIR_STATIC3,195,162,70,8
-    EDITTEXT        IDC_MACHEDPAIROPEN_EDIT3,267,160,14,14
-    EDITTEXT        IDC_MACHEDPAIRCLOSE_EDIT3,295,160,14,14
+    CONTROL         "Recall auto-completion on each input",IDD_AUTOC_RECALL_CHECK,"Button",BS_AUTOCHECKBOX | WS_TABSTOP,38,83,160,10
```
[the remark this answers is not in the document]
Comment (mpheath) [reply PowerEditor/src/WinControls/Preference/preference.rc]: 1. Renamed to "Make auto-completion list brief".
 2. Removed "Filter the list items".
 
The option of "Function completion" to be filtered without brief is now not available.

**Comment by donho** [comment]
@ArkadiuszMichalski 
> This allows us to manually select those that can't be selected, such as #include.

Hmm... I think "#include" keyword is another problem independently here, and it deserves an issue for that.

File: PowerEditor/src/ScintillaComponent/AutoCompletion.cpp [hunk 2]
```diff
@@ -268,7 +268,7 @@ static bool isAllDigits(const generic_string &str)
 	return true;
 }
 
-void sortInsensitive(vector<generic_string> &wordArray)
+static void sortInsensitive(vector<generic_string> &wordArray)
```
Comment (donho) [inline PowerEditor/src/ScintillaComponent/AutoCompletion.cpp:271]: Why should it be static ?

**Review by donho — APPROVED** [review]

**Comment by ArkadiuszMichalski** [comment]
@donho 
>Hmm... I think "#include" keyword is another problem independently here, and it deserves an issue for that.

Yes, I suggested same here https://github.com/notepad-plus-plus/notepad-plus-plus/pull/13075#issuecomment-1425418483.

I test this one again and back with one question: why `Function completion` doesn't make filter, same as `Function and world completion` when suggestion is triggered automatically?

If we need to view all available entries from the .xml file, we can do it by manually trigger the hint (using a shortcut). This works in both cases for `Function completion` and `Function and world completion`.

**Comment by donho** [comment]
> I test this one again and back with one question: why `Function completion` doesn't make filter, same as `Function and world completion` when suggestion is triggered automatically?

Good question.
It's just because they are done with the different time, and IMO the whole list is not so inconvenient (with automatic jumping) so users don't really care. Do you guys think it's better to have the filter on function list? Besides to align the behaviours, what's the advantage?

**Comment by ArkadiuszMichalski** [comment]
For me personally, it doesn't matter. Since no one has complained about it so far, it's better not to change it. It's good that for `Function and world completion` we can call the list manually and view all items (without having to look into the XML file). That's enough.
If needed, it would probably be best to add a switch `[ ] Filter the list items`, which would work for both `Function completion` and `Function and world completion` (so it would be consistent), or even make separate switches for both mechanisms.

It's better to focus on the problem with `#include` because I've seen that a lot of xml files use special characters in words and of course they won't be selected (without combining).

**Comment by Yaron10** [comment]
@donho @mpheath,

**STR:**
Uncheck `Enable auto-completion on each input`.
Restart NPP.

**Result:**
`Make auto-completion list brief` is wrongly enabled.

**Comment by donho** [comment]
@Yaron10 
Thank you
Fixed in https://github.com/notepad-plus-plus/notepad-plus-plus/commit/52c850de74929206814c1e9edc5db4c275e9fc45

**Comment by Yaron10** [comment]
@donho,

Great. 👍 
Thank you.

**Pull request closed without merging**